File: src/core/repair/pianoRepairSolver.ts
import type {
    SectionArtifactSummary,
    SectionRenderEventArtifact,
} from "../pipeline/types.js";
import { applyPianoProjection } from "./pianoProjection.js";

// pianoRepairSolver.ts — Piano idiom repair pass
// ──────────────────────────────────────────────────────────────────────────────
// Sits between NotaGen keyboard candidates and the craft-scoring gate.
// Post-processes SectionArtifactSummary event arrays to correct piano-idiom
// violations without re-composing the piece.  Musical contour, rhythm, and
// harmonic intent are preserved; only physical realization is adjusted.
//
// Pipeline position:
//   NotaGen candidates → ABC/MIDI projection → piano_repair_solver.py (Python)
//   → piano playability gate → PianoRepairSolver (this, TS) → craft scoring
//   → localized rewrite
//
// Relationship to Python repair (piano_repair_solver.py):
//   Python repair runs BEFORE the MIDI file is written and directly corrects
//   the rendered audio.  This module runs AFTERWARDS on SectionArtifactSummary
//   event arrays to re-derive the 21 piano* evidence fields so craft scoring
//   and Gate 3 see accurate post-repair metrics.  When `proposalMidiRewritten`
//   is true in the Python response the MIDI already reflects all corrections;
//   this module's job is then purely to update scoring evidence.
//
// Seven repair kinds:
//   1. chord_span_revoice    — inner voices dropped when hand span exceeds limit
//   2. register_correction   — notes outside idiomatic range octave-shifted
//   3. leap_attenuation      — RH melody leaps > limit compressed by ±12
//   4. bass_reinforcement    — LH lacking bass territory shifted into range
//   5. voicing_clarity       — LH notes crowding RH register removed
//   6. pedal_change_increase — blur risk metadata flagged; change count updated
//   7. chord_thinning        — chords exceeding voice-count limit trimmed
//
// The module is pure / immutable: original artifacts are never mutated.
// After all repairs, applyPianoProjection() re-derives the 21 projection
// evidence fields so downstream evaluators see updated metrics.
// ──────────────────────────────────────────────────────────────────────────────

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

export type PianoRepairKind =
    | "chord_span_revoice"
    | "register_correction"
    | "leap_attenuation"
    | "bass_reinforcement"
    | "voicing_clarity"
    | "pedal_change_increase"
    | "chord_thinning";

export interface PianoRepairAction {
    kind: PianoRepairKind;
    sectionId: string;
    description: string;
    severity: "minor" | "moderate" | "major";
}

export interface PianoRepairOptions {
    /** Max comfortable right-hand chord span in semitones (default 14 = major 9th). */
    maxRHSpan?: number;
    /** Max comfortable left-hand chord span in semitones (default 12 = octave). */
    maxLHSpan?: number;
    /** Minimum idiomatic RH pitch — below this is shifted up (default 48 = C3). */
    rhRegisterFloor?: number;
    /** Maximum idiomatic LH pitch — above this is shifted down (default 72 = C5). */
    lhRegisterCeiling?: number;
    /** Maximum melodic interval before attenuation kicks in, semitones (default 12). */
    maxLeapSemitones?: number;
    /** Maximum simultaneous note count per chord event (default 6). */
    maxSimultaneousNotes?: number;
    /** pianoPedalBlurRisk threshold above which pedal changes are recommended (default 0.60). */
    pedalBlurThreshold?: number;
    /** Restrict to only these repair kinds.  Omit or pass undefined to enable all. */
    enabledRepairs?: PianoRepairKind[];
}

export interface PianoRepairResult {
    sectionId: string;
    /** true when at least one repair was applied. */
    repaired: boolean;
    repairCount: number;
    actions: PianoRepairAction[];
    /** A new SectionArtifactSummary with updated events and projection evidence. */
    updatedArtifact: SectionArtifactSummary;
    /** True when the Python pipeline already rewrote the MIDI file via
     *  piano_repair_solver.py + write_midi_from_events().  In that case this
     *  TypeScript pass updates scoring evidence only; the rendered audio is
     *  already corrected.  Callers should set this from `proposalMidiRewritten`
     *  in the LearnedSymbolicProposalResponse. */
    midiRewritten?: boolean;
}

// ---------------------------------------------------------------------------
// Internal option resolution
// ---------------------------------------------------------------------------

const ALL_REPAIRS: PianoRepairKind[] = [
    "chord_span_revoice",
    "register_correction",
    "leap_attenuation",
    "bass_reinforcement",
    "voicing_clarity",
    "pedal_change_increase",
    "chord_thinning",
];

interface ResolvedOptions {
    maxRHSpan: number;
    maxLHSpan: number;
    rhRegisterFloor: number;
    lhRegisterCeiling: number;
    maxLeapSemitones: number;
    maxSimultaneousNotes: number;
    pedalBlurThreshold: number;
    enabledRepairs: Set<PianoRepairKind>;
}

function resolveOptions(opts?: PianoRepairOptions): ResolvedOptions {
    return {
        maxRHSpan:            opts?.maxRHSpan            ?? 14,
        maxLHSpan:            opts?.maxLHSpan            ?? 12,
        rhRegisterFloor:      opts?.rhRegisterFloor      ?? 48,
        lhRegisterCeiling:    opts?.lhRegisterCeiling    ?? 72,
        maxLeapSemitones:     opts?.maxLeapSemitones     ?? 12,
        maxSimultaneousNotes: opts?.maxSimultaneousNotes ?? 6,
        pedalBlurThreshold:   opts?.pedalBlurThreshold   ?? 0.60,
        enabledRepairs:       new Set(opts?.enabledRepairs ?? ALL_REPAIRS),
    };
}

// ---------------------------------------------------------------------------
// Event helpers (immutable)
// ---------------------------------------------------------------------------

function pitchesOf(event: SectionRenderEventArtifact): number[] {
    if (event.type === "chord") {
        if (event.pitches?.length) return event.pitches;
        if (event.pitch !== undefined) return [event.pitch];
    }
    if (event.type === "note" && event.pitch !== undefined) return [event.pitch];
    return [];
}

function withPitches(
    event: SectionRenderEventArtifact,
    newPitches: number[],
): SectionRenderEventArtifact {
    if (newPitches.length === 0) return event;
    if (newPitches.length === 1) {
        return { ...event, type: "note", pitch: newPitches[0], pitches: undefined };
    }
    return { ...event, type: "chord", pitches: newPitches, pitch: undefined };
}

// ---------------------------------------------------------------------------
// Repair 1: chord_span_revoice
// ---------------------------------------------------------------------------

/**
 * When a chord event's span (max − min pitch) exceeds maxSpan, inner voices
 * are pruned until the span fits.  The bass (lowest) and melody (highest)
 * notes are always kept.
 */
function repairChordSpan(
    events: SectionRenderEventArtifact[],
    maxSpan: number,
    sectionId: string,
): { events: SectionRenderEventArtifact[]; actions: PianoRepairAction[] } {
    const actions: PianoRepairAction[] = [];
    let count = 0;

    const out = events.map((ev) => {
        const ps = pitchesOf(ev);
        if (ps.length < 2) return ev;

        const span = Math.max(...ps) - Math.min(...ps);
        if (span <= maxSpan) return ev;

        const sorted = [...ps].sort((a, b) => a - b);
        const bass   = sorted[0]!;
        const melody = sorted[sorted.length - 1]!;
        const inner  = sorted.slice(1, -1).filter((p) => p - bass <= maxSpan);
        const newPs  = [bass, ...inner, melody];
        if (newPs.length === ps.length) return ev;

        count++;
        return withPitches(ev, newPs);
    });

    if (count > 0) {
        actions.push({
            kind: "chord_span_revoice",
            sectionId,
            description: `re-voiced ${count} chord(s) exceeding span of ${maxSpan} semitones`,
            severity: count > 4 ? "major" : "moderate",
        });
    }
    return { events: out, actions };
}

// ---------------------------------------------------------------------------
// Repair 2: register_correction
// ---------------------------------------------------------------------------

/**
 * Notes outside the idiomatic register are shifted by ±12 semitones.
 * Right hand: below `floor` → +12.  Left hand: above `ceiling` → −12.
 */
function repairRegister(
    events: SectionRenderEventArtifact[],
    floor: number,
    ceiling: number,
    sectionId: string,
    hand: "right" | "left",
): { events: SectionRenderEventArtifact[]; actions: PianoRepairAction[] } {
    const actions: PianoRepairAction[] = [];
    let tooLow = 0;
    let tooHigh = 0;

    const out = events.map((ev) => {
        const ps = pitchesOf(ev);
        if (ps.length === 0) return ev;

        const adjusted = ps.map((p) => {
            if (p < floor)   { tooLow++;  return p + 12; }
            if (p > ceiling) { tooHigh++; return p - 12; }
            return p;
        });
        if (adjusted.every((p, i) => p === ps[i])) return ev;
        return withPitches(ev, adjusted);
    });

    const label = hand === "right" ? "RH" : "LH";
    if (tooLow > 0) {
        actions.push({
            kind: "register_correction",
            sectionId,
            description: `${label}: +12 on ${tooLow} note(s) below MIDI ${floor}`,
            severity: tooLow > 4 ? "moderate" : "minor",
        });
    }
    if (tooHigh > 0) {
        actions.push({
            kind: "register_correction",
            sectionId,
            description: `${label}: −12 on ${tooHigh} note(s) above MIDI ${ceiling}`,
            severity: tooHigh > 4 ? "moderate" : "minor",
        });
    }
    return { events: out, actions };
}

// ---------------------------------------------------------------------------
// Repair 3: leap_attenuation  (melody / right-hand single notes only)
// ---------------------------------------------------------------------------

/**
 * When two consecutive single-note melody events form an interval larger than
 * maxLeap semitones, the second note is octave-transposed toward the first to
 * minimise the leap.  Chord events and rest events are skipped.
 */
function repairLeaps(
    events: SectionRenderEventArtifact[],
    maxLeap: number,
    sectionId: string,
): { events: SectionRenderEventArtifact[]; actions: PianoRepairAction[] } {
    const actions: PianoRepairAction[] = [];
    const out = events.map((e) => ({ ...e })); // shallow copy

    let repaired = 0;
    let prevPitch: number | undefined;

    for (let i = 0; i < out.length; i++) {
        const ev = out[i]!;
        if (ev.type === "rest") { prevPitch = undefined; continue; }
        if (ev.type !== "note" || ev.pitch === undefined) {
            // For chord events, update prevPitch to the top note
            const ps = pitchesOf(ev);
            if (ps.length > 0) prevPitch = Math.max(...ps);
            continue;
        }

        if (prevPitch !== undefined) {
            const interval = ev.pitch - prevPitch;
            if (Math.abs(interval) > maxLeap) {
                const shift = interval > 0 ? -12 : 12;
                out[i] = { ...ev, pitch: ev.pitch + shift };
                repaired++;
            }
        }
        prevPitch = out[i]!.pitch;
    }

    if (repaired > 0) {
        actions.push({
            kind: "leap_attenuation",
            sectionId,
            description: `attenuated ${repaired} leap(s) > ${maxLeap} semitones`,
            severity: repaired > 3 ? "moderate" : "minor",
        });
    }
    return { events: out, actions };
}

// ---------------------------------------------------------------------------
// Repair 4: bass_reinforcement
// ---------------------------------------------------------------------------

// Notes at or below this MIDI pitch are considered in "bass territory".
const BASS_TERRITORY_CEIL = 52; // E3

/**
 * When no left-hand event contains a note below E3 (MIDI 52), the section
 * has no bass grounding.  The globally lowest-pitched events are shifted down
 * one octave to put them in bass territory.
 */
function repairBassReinforcement(
    events: SectionRenderEventArtifact[],
    sectionId: string,
): { events: SectionRenderEventArtifact[]; actions: PianoRepairAction[] } {
    const actions: PianoRepairAction[] = [];
    const noteEvents = events.filter((e) => e.type !== "rest");
    if (noteEvents.length === 0) return { events, actions };

    const allPitches = noteEvents.flatMap(pitchesOf);
    if (allPitches.length === 0) return { events, actions };

    const hasBass = allPitches.some((p) => p <= BASS_TERRITORY_CEIL);
    if (hasBass) return { events, actions };

    const globalMin = Math.min(...allPitches);
    let shifted = 0;

    const out = events.map((ev) => {
        const ps = pitchesOf(ev);
        if (ps.length === 0) return ev;
        if (!ps.includes(globalMin)) return ev;

        const adjusted = ps.map((p) => (p === globalMin ? p - 12 : p));
        shifted++;
        return withPitches(ev, adjusted);
    });

    if (shifted > 0) {
        actions.push({
            kind: "bass_reinforcement",
            sectionId,
            description: `LH lacks bass below E3 — shifted ${shifted} bass event(s) down one octave from MIDI ${globalMin}`,
            severity: "moderate",
        });
    }
    return { events: out, actions };
}

// ---------------------------------------------------------------------------
// Repair 5: voicing_clarity
// ---------------------------------------------------------------------------

/**
 * Left-hand notes that intrude above the right-hand's lowest sounding pitch
 * collide with the melody register.  Such notes are removed from the LH event
 * to clear space for the melody.  Events that would become empty are left
 * unchanged to avoid silencing an entire beat.
 */
function repairVoicingClarity(
    lhEvents: SectionRenderEventArtifact[],
    rhMinPitch: number,
    sectionId: string,
): { events: SectionRenderEventArtifact[]; actions: PianoRepairAction[] } {
    const actions: PianoRepairAction[] = [];
    if (rhMinPitch <= 0) return { events: lhEvents, actions };

    let removed = 0;

    const out = lhEvents.map((ev) => {
        const ps = pitchesOf(ev);
        if (ps.length === 0) return ev;

        const filtered = ps.filter((p) => p < rhMinPitch);
        if (filtered.length === ps.length) return ev;
        if (filtered.length === 0) return ev; // keep rather than silence
        removed += ps.length - filtered.length;
        return withPitches(ev, filtered);
    });

    if (removed > 0) {
        actions.push({
            kind: "voicing_clarity",
            sectionId,
            description: `removed ${removed} LH note(s) at or above RH floor (MIDI ${rhMinPitch}) to clear melody space`,
            severity: removed > 5 ? "moderate" : "minor",
        });
    }
    return { events: out, actions };
}

// ---------------------------------------------------------------------------
// Repair 6: pedal_change_increase  (metadata only — no event synthesis)
// ---------------------------------------------------------------------------

/**
 * When pianoPedalBlurRisk exceeds the threshold, sustained pedal is blurring
 * harmony changes.  This repair increases the `pianoPedalChangeCount` field
 * on the artifact and logs the recommendation; actual MIDI pedal events are
 * handled downstream by the humanizer.
 */
function repairPedalChangeIncrease(
    artifact: SectionArtifactSummary,
    threshold: number,
): { actions: PianoRepairAction[]; additionalPedalChanges: number } {
    const blurRisk = artifact.pianoPedalBlurRisk ?? 0;
    if (blurRisk <= threshold) return { actions: [], additionalPedalChanges: 0 };

    const additionalChanges = Math.max(1, Math.ceil(artifact.measureCount * (blurRisk - threshold) * 2));
    return {
        actions: [{
            kind: "pedal_change_increase",
            sectionId: artifact.sectionId,
            description: `blur risk ${blurRisk.toFixed(2)} > threshold ${threshold}: +${additionalChanges} pedal change(s) recommended`,
            severity: blurRisk > 0.80 ? "major" : "moderate",
        }],
        additionalPedalChanges: additionalChanges,
    };
}

// ---------------------------------------------------------------------------
// Repair 7: chord_thinning
// ---------------------------------------------------------------------------

/**
 * Chord events with more simultaneous notes than maxNotes are trimmed.
 * The bass (lowest) and melody (highest) pitches are always preserved.
 * Inner voices are kept from the top down to fill the allowed count.
 */
function repairChordThinning(
    events: SectionRenderEventArtifact[],
    maxNotes: number,
    sectionId: string,
): { events: SectionRenderEventArtifact[]; actions: PianoRepairAction[] } {
    const actions: PianoRepairAction[] = [];
    let count = 0;

    const out = events.map((ev) => {
        const ps = pitchesOf(ev);
        if (ps.length <= maxNotes) return ev;

        const sorted  = [...ps].sort((a, b) => a - b);
        const bass    = sorted[0]!;
        const melody  = sorted[sorted.length - 1]!;
        const inner   = sorted.slice(1, -1);
        const keepInner = inner.slice(-(maxNotes - 2));
        const newPs   = [bass, ...keepInner, melody];

        count++;
        return withPitches(ev, newPs);
    });

    if (count > 0) {
        actions.push({
            kind: "chord_thinning",
            sectionId,
            description: `thinned ${count} chord(s) to max ${maxNotes} simultaneous notes`,
            severity: count > 3 ? "moderate" : "minor",
        });
    }
    return { events: out, actions };
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Applies all enabled piano-idiom repairs to a single section artifact.
 *
 * Repair order:
 *   1. chord_span_revoice    — structural: fix wide spans first
 *   2. register_correction   — pitch range: fix out-of-range notes
 *   3. leap_attenuation      — melodic shape: compress large leaps
 *   4. bass_reinforcement    — LH bass: ensure bass territory coverage
 *   5. voicing_clarity       — register separation: clear melody space
 *   6. pedal_change_increase — pedal metadata update
 *   7. chord_thinning        — density: final voice-count pass
 *
 * After repairs, `applyPianoProjection()` re-derives the 21 flat piano*
 * evidence fields so Gate 3 and craft scoring see accurate metrics.
 */
export function repairPianoSection(
    artifact: SectionArtifactSummary,
    options?: PianoRepairOptions,
): PianoRepairResult {
    const opts = resolveOptions(options);
    const { sectionId } = artifact;
    const allActions: PianoRepairAction[] = [];

    let rhEvents = [...artifact.melodyEvents];
    let lhEvents = [...artifact.accompanimentEvents];

    // 1. chord_span_revoice
    if (opts.enabledRepairs.has("chord_span_revoice")) {
        const r1rh = repairChordSpan(rhEvents, opts.maxRHSpan, sectionId);
        const r1lh = repairChordSpan(lhEvents, opts.maxLHSpan, sectionId);
        rhEvents = r1rh.events;
        lhEvents = r1lh.events;
        allActions.push(...r1rh.actions, ...r1lh.actions);
    }

    // 2. register_correction
    if (opts.enabledRepairs.has("register_correction")) {
        const r2rh = repairRegister(rhEvents, opts.rhRegisterFloor, 108, sectionId, "right");
        const r2lh = repairRegister(lhEvents, 24, opts.lhRegisterCeiling, sectionId, "left");
        rhEvents = r2rh.events;
        lhEvents = r2lh.events;
        allActions.push(...r2rh.actions, ...r2lh.actions);
    }

    // 3. leap_attenuation (RH melody only)
    if (opts.enabledRepairs.has("leap_attenuation")) {
        const r3 = repairLeaps(rhEvents, opts.maxLeapSemitones, sectionId);
        rhEvents = r3.events;
        allActions.push(...r3.actions);
    }

    // 4. bass_reinforcement (LH)
    if (opts.enabledRepairs.has("bass_reinforcement")) {
        const r4 = repairBassReinforcement(lhEvents, sectionId);
        lhEvents = r4.events;
        allActions.push(...r4.actions);
    }

    // 5. voicing_clarity (LH, using current RH min pitch as boundary)
    if (opts.enabledRepairs.has("voicing_clarity")) {
        const rhPitches = rhEvents.flatMap(pitchesOf);
        const rhMinPitch = rhPitches.length > 0 ? Math.min(...rhPitches) : 0;
        const r5 = repairVoicingClarity(lhEvents, rhMinPitch, sectionId);
        lhEvents = r5.events;
        allActions.push(...r5.actions);
    }

    // 6. pedal_change_increase (metadata)
    let additionalPedalChanges = 0;
    if (opts.enabledRepairs.has("pedal_change_increase")) {
        const r6 = repairPedalChangeIncrease(artifact, opts.pedalBlurThreshold);
        additionalPedalChanges = r6.additionalPedalChanges;
        allActions.push(...r6.actions);
    }

    // 7. chord_thinning (both hands)
    if (opts.enabledRepairs.has("chord_thinning")) {
        const r7rh = repairChordThinning(rhEvents, opts.maxSimultaneousNotes, sectionId);
        const r7lh = repairChordThinning(lhEvents, opts.maxSimultaneousNotes, sectionId);
        rhEvents = r7rh.events;
        lhEvents = r7lh.events;
        allActions.push(...r7rh.actions, ...r7lh.actions);
    }

    const patchedEvents: SectionArtifactSummary = {
        ...artifact,
        melodyEvents:        rhEvents,
        accompanimentEvents: lhEvents,
    };

    // Re-derive the 21 piano* projection evidence fields after event repairs.
    // Projection overwrites pianoPedalChangeCount with layout.pedalEventCount,
    // so the pedal-change recommendation is applied AFTER projection.
    let updatedArtifact = applyPianoProjection(patchedEvents);

    if (additionalPedalChanges > 0) {
        updatedArtifact = {
            ...updatedArtifact,
            pianoPedalChangeCount: (updatedArtifact.pianoPedalChangeCount ?? 0) + additionalPedalChanges,
        };
    }

    return {
        sectionId,
        repaired:    allActions.length > 0,
        repairCount: allActions.length,
        actions:     allActions,
        updatedArtifact,
    };
}

/**
 * Batch version of repairPianoSection.
 * Repairs all sections and returns an aggregate result with the updated
 * artifact array ready for the next pipeline stage.
 */
export function repairPianoCandidates(
    artifacts: SectionArtifactSummary[],
    options?: PianoRepairOptions,
): {
    results:       PianoRepairResult[];
    anyRepaired:   boolean;
    repairedArtifacts: SectionArtifactSummary[];
} {
    const results = artifacts.map((a) => repairPianoSection(a, options));
    return {
        results,
        anyRepaired:       results.some((r) => r.repaired),
        repairedArtifacts: results.map((r) => r.updatedArtifact),
    };
}

File: workers/composer/learned_symbolic/piano_repair_solver.py
"""Piano idiom repair solver for the solo_piano_symbolic lane.

Runs in abc_project.py Stage 5.5, after piano enrichment and BEFORE the
optional MIDI re-write (Stage 6).  Repairs are therefore reflected in the
rendered MIDI output that listeners actually hear.

Mirrors the 7 repair kinds from src/pipeline/pianoRepairSolver.ts but operates
on Python event dicts produced by abc_to_events.py:

    {"kind": "note",  "midi": int, "quarterLength": float, "velocity": int, "role": str}
    {"kind": "chord", "midiPitches": [int], "quarterLength": float, "velocity": int, "role": str}
    {"kind": "rest",  "quarterLength": float, "role": str}

The TypeScript pianoRepairSolver.ts still runs afterwards on SectionArtifactSummary
event arrays for craft-score metric updates, but the underlying MIDI file is
already corrected here.  The `midi_rewritten` flag on AbcProjectionResult
signals to TypeScript that the MIDI reflects Python-side repairs.
"""

from __future__ import annotations

from typing import Any

Event = dict[str, Any]

# ─── Default repair thresholds ────────────────────────────────────────────────

MAX_RH_SPAN: int = 14   # comfortable RH chord span, semitones (major 9th)
MAX_LH_SPAN: int = 12   # comfortable LH chord span, semitones (octave)
RH_FLOOR: int = 48      # C3 — RH notes below here are shifted up
LH_CEILING: int = 72    # C5 — LH notes above here are shifted down
MAX_LEAP: int = 12      # RH melodic leap limit before attenuation (semitones)
BASS_CEILING: int = 52  # E3 — pitches below this are "bass territory"
MAX_NOTES: int = 6      # max simultaneous voices per chord event


# ─── Helpers ─────────────────────────────────────────────────────────────────

def _pitches_of(event: Event) -> list[int]:
    """Return all MIDI pitches in an event (empty list for rests)."""
    if event.get("kind") == "chord":
        raw = event.get("midiPitches") or []
        return [int(p) for p in raw if isinstance(p, (int, float))]
    p = event.get("midi") or event.get("pitch")
    if isinstance(p, (int, float)):
        return [int(p)]
    return []


def _set_pitches(event: Event, pitches: list[int]) -> Event:
    """Return a shallow copy of *event* with updated pitch data."""
    ev = dict(event)
    if ev.get("kind") == "chord":
        ev["midiPitches"] = sorted(pitches)
    else:
        ev["midi"] = pitches[0] if pitches else int(event.get("midi") or 60)
    return ev


# ─── Repair 1: chord_span_revoice ────────────────────────────────────────────

def _repair_chord_span(
    events: list[Event],
    max_span: int,
    label: str,
) -> tuple[list[Event], list[dict[str, Any]]]:
    """Drop inner voices from chord events whose span exceeds *max_span*.

    Bass (lowest) and melody (highest) pitches are always preserved.
    Inner voices are greedily added back (highest first) as long as the
    resulting span stays within *max_span*.
    """
    out: list[Event] = []
    actions: list[dict[str, Any]] = []

    for ev in events:
        if ev.get("kind") != "chord":
            out.append(ev)
            continue

        pitches = _pitches_of(ev)
        if len(pitches) < 2 or (max(pitches) - min(pitches)) <= max_span:
            out.append(ev)
            continue

        keep = [pitches[0], pitches[-1]]
        for p in reversed(pitches[1:-1]):
            candidate = sorted(keep + [p])
            if max(candidate) - min(candidate) <= max_span:
                keep.append(p)
        keep = sorted(keep)

        if keep != pitches:
            out.append(_set_pitches(ev, keep))
            actions.append({
                "kind": "chord_span_revoice",
                "label": label,
                "before": pitches,
                "after": keep,
            })
        else:
            out.append(ev)

    return out, actions


# ─── Repair 2: register_correction ───────────────────────────────────────────

def _repair_register(
    events: list[Event],
    floor: int,
    ceiling: int,
    label: str,
) -> tuple[list[Event], list[dict[str, Any]]]:
    """Octave-shift notes that fall outside [floor, ceiling]."""
    out: list[Event] = []
    actions: list[dict[str, Any]] = []

    for ev in events:
        pitches = _pitches_of(ev)
        if not pitches:
            out.append(ev)
            continue

        new_pitches = list(pitches)
        changed = False
        for i, p in enumerate(new_pitches):
            while new_pitches[i] < floor:
                new_pitches[i] += 12
                changed = True
            while new_pitches[i] > ceiling:
                new_pitches[i] -= 12
                changed = True

        if changed:
            out.append(_set_pitches(ev, new_pitches))
            actions.append({
                "kind": "register_correction",
                "label": label,
                "before": pitches,
                "after": new_pitches,
            })
        else:
            out.append(ev)

    return out, actions


# ─── Repair 3: leap_attenuation ──────────────────────────────────────────────

def _repair_leaps(
    events: list[Event],
    max_leap: int,
) -> tuple[list[Event], list[dict[str, Any]]]:
    """Compress RH melody leaps > *max_leap* semitones by ±12 toward previous."""
    note_positions = [
        (i, ev) for i, ev in enumerate(events) if ev.get("kind") == "note"
    ]
    modified: dict[int, int] = {}
    actions: list[dict[str, Any]] = []

    for k in range(1, len(note_positions)):
        prev_idx, prev_ev = note_positions[k - 1]
        curr_idx, curr_ev = note_positions[k]
        prev_pitch = modified.get(prev_idx, int(prev_ev.get("midi") or 60))
        curr_pitch = int(curr_ev.get("midi") or 60)
        interval = curr_pitch - prev_pitch

        if abs(interval) > max_leap:
            adjustment = -12 if interval > 0 else 12
            modified[curr_idx] = curr_pitch + adjustment
            actions.append({
                "kind": "leap_attenuation",
                "before": curr_pitch,
                "after": modified[curr_idx],
                "interval": interval,
            })

    if not modified:
        return events, []

    out = list(events)
    for idx, new_pitch in modified.items():
        ev = dict(events[idx])
        ev["midi"] = new_pitch
        out[idx] = ev
    return out, actions


# ─── Repair 4: bass_reinforcement ────────────────────────────────────────────

def _repair_bass_reinforcement(
    events: list[Event],
) -> tuple[list[Event], list[dict[str, Any]]]:
    """If no LH event has a pitch in bass territory, shift the lowest down 12."""
    all_pitches = [p for ev in events for p in _pitches_of(ev)]
    if not all_pitches or min(all_pitches) < BASS_CEILING:
        return events, []

    global_min = min(all_pitches)
    out = list(events)
    for i, ev in enumerate(events):
        ps = _pitches_of(ev)
        if ps and min(ps) == global_min:
            new_ps = [p - 12 for p in ps]
            out[i] = _set_pitches(ev, new_ps)
            return out, [{
                "kind": "bass_reinforcement",
                "before": ps,
                "after": new_ps,
            }]

    return events, []


# ─── Repair 5: voicing_clarity ────────────────────────────────────────────────

def _repair_voicing_clarity(
    lh_events: list[Event],
    rh_min_pitch: int,
) -> tuple[list[Event], list[dict[str, Any]]]:
    """Remove LH notes that crowd the RH register (>= rh_min_pitch)."""
    out: list[Event] = []
    actions: list[dict[str, Any]] = []

    for ev in lh_events:
        if ev.get("kind") == "chord":
            pitches = _pitches_of(ev)
            kept = [p for p in pitches if p < rh_min_pitch]
            if not kept and pitches:
                kept = [pitches[0]]  # always preserve at least the bass note
            if len(kept) != len(pitches):
                out.append(_set_pitches(ev, kept))
                actions.append({
                    "kind": "voicing_clarity",
                    "before": pitches,
                    "after": kept,
                })
                continue
        elif ev.get("kind") == "note":
            p = int(ev.get("midi") or 0)
            if p >= rh_min_pitch:
                actions.append({"kind": "voicing_clarity", "removed_pitch": p})
                continue  # drop the note entirely
        out.append(ev)

    return out, actions


# ─── Repair 7: chord_thinning ────────────────────────────────────────────────

def _repair_chord_thinning(
    events: list[Event],
    max_notes: int,
) -> tuple[list[Event], list[dict[str, Any]]]:
    """Trim chords exceeding *max_notes* voices; always keep bass and melody."""
    out: list[Event] = []
    actions: list[dict[str, Any]] = []

    for ev in events:
        if ev.get("kind") != "chord":
            out.append(ev)
            continue

        pitches = _pitches_of(ev)
        if len(pitches) <= max_notes:
            out.append(ev)
            continue

        # Keep bass (pitches[0]) + top (max_notes-1) highest voices
        kept = sorted([pitches[0]] + pitches[-(max_notes - 1):])
        out.append(_set_pitches(ev, kept))
        actions.append({"kind": "chord_thinning", "before": pitches, "after": kept})

    return out, actions


# ─── Public API ───────────────────────────────────────────────────────────────

def repair_piano_sections(
    sections: list[dict[str, Any]],
    *,
    max_rh_span: int = MAX_RH_SPAN,
    max_lh_span: int = MAX_LH_SPAN,
    rh_floor: int = RH_FLOOR,
    lh_ceiling: int = LH_CEILING,
    max_leap: int = MAX_LEAP,
    max_notes: int = MAX_NOTES,
) -> tuple[list[dict[str, Any]], list[dict[str, Any]], bool]:
    """Apply all 7 piano repairs to sections enriched with rightHandEvents / leftHandEvents.

    Called after enrich_proposal_sections_with_piano_layout() so that each
    section already has rightHandEvents and leftHandEvents set.

    Repair 6 (pedal_change_increase) is metadata-only and handled at the
    AbcProjectionResult level, not here.

    Returns:
        (repaired_sections, repair_log, any_repaired)
        repair_log entries: {"sectionId": str, "actions": [{"kind": str, ...}]}
    """
    repaired_sections: list[dict[str, Any]] = []
    repair_log: list[dict[str, Any]] = []
    any_repaired = False

    for raw_section in sections:
        section = dict(raw_section)
        section_id = str(section.get("sectionId") or "")
        all_actions: list[dict[str, Any]] = []

        rh: list[Event] = list(section.get("rightHandEvents") or [])
        lh: list[Event] = list(section.get("leftHandEvents") or [])

        # 1. chord_span_revoice
        rh, a = _repair_chord_span(rh, max_rh_span, "RH")
        all_actions.extend(a)
        lh, a = _repair_chord_span(lh, max_lh_span, "LH")
        all_actions.extend(a)

        # 2. register_correction
        rh, a = _repair_register(rh, rh_floor, 108, "RH")
        all_actions.extend(a)
        lh, a = _repair_register(lh, 24, lh_ceiling, "LH")
        all_actions.extend(a)

        # 3. leap_attenuation (RH melody only)
        rh, a = _repair_leaps(rh, max_leap)
        all_actions.extend(a)

        # 4. bass_reinforcement (LH)
        lh, a = _repair_bass_reinforcement(lh)
        all_actions.extend(a)

        # 5. voicing_clarity (LH vs current RH floor)
        rh_pitches = [p for ev in rh for p in _pitches_of(ev)]
        rh_min = min(rh_pitches) if rh_pitches else 60
        lh, a = _repair_voicing_clarity(lh, rh_min)
        all_actions.extend(a)

        # 7. chord_thinning (both hands)
        rh, a = _repair_chord_thinning(rh, max_notes)
        all_actions.extend(a)
        lh, a = _repair_chord_thinning(lh, max_notes)
        all_actions.extend(a)

        section["rightHandEvents"] = rh
        section["leftHandEvents"] = lh

        if all_actions:
            any_repaired = True
            repair_log.append({"sectionId": section_id, "actions": all_actions})

        repaired_sections.append(section)

    return repaired_sections, repair_log, any_repaired
